Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "JSCBytecodeCacheVersion.h"
+
+#include <mach-o/dyld.h>
+#include <optional>
+#include <uuid/uuid.h>
+#include <wtf/DataLog.h>
+#include <wtf/spi/darwin/dyldSPI.h>
+#include <wtf/text/SuperFastHash.h>
+#include <wtf/text/WTFString.h>
+
+namespace JSC {
+
+namespace JSCBytecodeCacheVersionInternal {
+static constexpr bool verbose = false;
+}
+
+std::optional<uint32_t> computeJSCBytecodeCacheVersion()
+{
+    static std::optional<uint32_t> cacheVersion;
+
+    // Only compute if not already successfully computed.
+    if (!cacheVersion.has_value()) {
+        uuid_t uuid;
+        uuid_string_t uuidString = { };
+
+        uint32_t count = _dyld_image_count();
+        for (uint32_t i = 0; i < count; i++) {
+            const char* imageName = _dyld_get_image_name(i);
+            if (!imageName || String::fromUTF8(imageName).find("JavaScriptCore.framework"_s) == notFound)
+                continue;
+
+            dataLogLnIf(JSCBytecodeCacheVersionInternal::verbose, "i:", i, " imageName:", imageName);
+            const struct mach_header* header = _dyld_get_image_header(i);
+
+            if (header && _dyld_get_image_uuid(header, uuid)) {
```

Comment:
I don't think you need to search for the current framework.
I think, `_dyld_get_dlopen_image_header(RTLD_SELF)` should work (please double-check by looking at UUID v.s. framework UUID, and review dyld code).

Reply:
Hmm, but let's be a bit careful about this. Using `HAVE(DYLD_DLOPEN_IMAGE_HEADER_SPI)` is probably required.

Reply:
Or, I wonder if we should just use `dyld_image_header_containing_address`.

1. Wrap two functions, each one has `NEVER_INLINE` and `JS_EXPORT_PRIVATE`
2. Both should use NO_TAIL_CALLS to suppress tail calls 
3. And in the nested function, using `__builtin_return_address(0)` to obtain the outer function's address. Then this address is 100% JSC framework's address
4. Pass it to `dyld_image_header_containing_address`

Reply:
Couldn't we just call `dyld_image_header_containing_address(computeJSCBytecodeCacheVersion)`? Even if it gets inlined that function should still by in the dylib, IIUC.